pkg/istio-agent/xds-proxy.go

```diff
@@ -0,0 +1,329 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istioagent
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"golang.org/x/oauth2"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/reflection"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultClientMaxReceiveMessageSize = 1024 * 1024 * 8
+	defaultInitialConnWindowSize       = 1024 * 1024 // default gRPC InitialWindowSize
+	defaultInitialWindowSize           = 1024 * 1024 // default gRPC ConnWindowSize
+)
+
+type XdsProxy struct {
+	stopChan             chan struct{}
+	downstreamListener   net.Listener
+	downstreamGrpcServer *grpc.Server
+	upstreamAddress      string
+	upstreamDialOptions  []grpc.DialOption
+}
+
+// XDS Proxy proxies all XDS requests from envoy to istiod, in addition to allowing subsystems inside
+// the agent to also communicate with either istiod/envoy (eg dns, sds, etc).
+// The goal here is to consolidate all xds related connections to istiod/envoy into a
+// single tcp connection with multiple gRPC streams.
+// TODO: Right now, the workloadSDS server and gatewaySDS servers are still separate
+// connections. These need to be consolidated
+func initXdsProxy(sa *Agent) (*XdsProxy, error) {
+	var err error
+	proxy := &XdsProxy{
+		upstreamAddress: sa.proxyConfig.DiscoveryAddress,
+	}
+
+	if err = proxy.initDownstreamServer(); err != nil {
+		return nil, err
+	}
+
+	if proxy.upstreamDialOptions, err = buildUpstreamClientDialOpts(sa); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		_ = proxy.downstreamGrpcServer.Serve(proxy.downstreamListener)
+	}()
+
+	return proxy, nil
+}
+
+// Every time envoy makes a fresh connection to the agent, we reestablish a new connection to the upstream xds
+// This ensures that a new connection between istiod and agent doesn't end up consuming pending messages from envoy
+// as the new connection may not go to the same istiod. Vice versa case also applies.
```

Comment:
Do we need a new connection? or just a new stream?

Reply:
grpc doesn't let us control the connections. If we try to open a new stream and the underlying connection is dead, grpc will try to reconnect (or so says the interweb :) ).